walnuts_cpp/test_stan.cpp

```diff
@@ -0,0 +1,203 @@
+                               std::string(dlerror()));
+    }
+
+    char *err;
+    model_ptr = model_construct(data, seed, &err);
+    if (!model_ptr) {
+      if (err) {
+        std::string error_string(err);
+        free_error_msg(err);
+        throw std::runtime_error(error_string);
+      }
+      throw std::runtime_error("Failed to construct model");
+    }
+  }
+
+  DynamicStanModel(const DynamicStanModel &) = delete; // non-copyable
+  DynamicStanModel &
+  operator=(const DynamicStanModel &) = delete; // non-copyable
+
+  DynamicStanModel(DynamicStanModel &&other)
+      : handle_(other.handle_), model_ptr(other.model_ptr),
+        model_construct(other.model_construct),
+        free_error_msg(other.free_error_msg),
+        model_destruct(other.model_destruct),
+        param_unc_num(other.param_unc_num),
+        log_density_gradient(other.log_density_gradient) {}
+
+  DynamicStanModel &operator=(DynamicStanModel &&other) {
+    if (this != &other) {
+      handle_ = other.handle_;
+      model_ptr = other.model_ptr;
+      model_construct = other.model_construct;
+      free_error_msg = other.free_error_msg;
+      model_destruct = other.model_destruct;
+      param_unc_num = other.param_unc_num;
+      log_density_gradient = other.log_density_gradient;
+
+      other.handle_ = nullptr;
+      other.model_ptr = nullptr;
+    }
+    return *this;
+  }
+
+  ~DynamicStanModel() {
+    if (handle_) {
+      if (model_ptr) {
+        model_destruct(model_ptr);
+      }
+      dlclose(handle_);
+    }
+  }
+
+  int size() const { return param_unc_num(model_ptr); }
+
+  void logp_grad(const Eigen::Matrix<double, Eigen::Dynamic, 1> &x,
+                 double &logp,
+                 Eigen::Matrix<double, Eigen::Dynamic, 1> &grad) const {
+
+    grad.resizeLike(x);
+
+    char *err;
+    auto start = std::chrono::high_resolution_clock::now();
+    int ret = log_density_gradient(model_ptr, true, true, x.data(), &logp,
+                                   grad.data(), &err);
+    auto end = std::chrono::high_resolution_clock::now();
+    total_time += std::chrono::duration<double>(end - start).count();
+    ++count;
+
+    if (ret != 0) {
+      if (err) {
+        std::string error_string(err);
+        free_error_msg(err);
+        throw std::runtime_error(error_string);
+      }
+      throw std::runtime_error("Failed to compute log density and gradient");
+    }
+  }
+
+private:
+  void *handle_;
+  bs_model *model_ptr;
+  decltype(&bs_model_construct) model_construct;
+  decltype(&bs_free_error_msg) free_error_msg;
+  decltype(&bs_model_destruct) model_destruct;
+  decltype(&bs_param_unc_num) param_unc_num;
+  decltype(&bs_log_density_gradient) log_density_gradient;
+};
+
+int main(int argc, char *argv[]) {
+  int seed = 333456;
+  int N = 10000;
+  double step_size = 0.025;
+  int max_depth = 10;
+
+  char *lib;
+  char *data;
+
+  // require at least the library name
+  if (argc > 2) {
```

Comment:
[optional]

I like to handle the exceptional cases first as a matter of style, which here would be

```cpp
if (argc <= 1) {  
  ... report err and return 1 ...
}
```